### PR TITLE
luci-proto-openfortivpn: cert/key PEM support, persistent option, SHA label fix

### PR DESCRIPTION
**Pull request by clayface**
Linked packages PR: [#16902](https://github.com/openwrt/packages/pull/16902)

Add support for user cert, key and CA cert in luci. This is largely replicated from luci-proto-anyconnect.

Add support for persistent option, to reconnect after timeout (default 5 minutes)

Finally correct the trusted cert SHA label from SHA1 to SHA256.

Signed-off-by: Matthew Hagan <mnhagan88@gmail.com>

### CODE REVIEW

**Comment by clayface** [comment]
ping @aaronjg

File: protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+end
+
+if arg[1] == "list" then
+	print(json.stringify({
+		getCertificates = {
+			interface = "interface"
+		},
+		setCertificates = {
+			interface = "interface",
+			user_cert = "PEM file data",
```
Comment (feckert) [inline protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn:44]: I think that the names for the ubus variables are not so well chosen. From my point of view, it should be a word without a space.

user_cert = "user-cert"
user_key = "user-key"
ca_file = "ca-file"

File: protocols/luci-proto-openfortivpn/htdocs/luci-static/resources/protocol/openfortivpn.js [hunk 2]
```diff
@@ -52,12 +106,53 @@ return network.registerProtocol('openfortivpn', {
...
+			return callSetCertificateFiles(section_id, null, null, sanitizeCert(value));
+		};
+
 		o = s.taboption('advanced', widgets.NetworkSelect, 'tunlink', _('Bind interface'), _('Bind the tunnel to this interface (optional).'));
 		o.exclude = s.section;
 		o.nocreate = true;
 		o.optional = true;
 
-		o = s.taboption('advanced', form.Value, 'trusted_cert', _("VPN Server's certificate SHA1 hash"));
+		o = s.taboption('advanced', form.Value, 'persist_int', _('Persistent reconnect interval'),_("Optional, in seconds. If set to '0', no reconnect is attempted."));
```
Comment (feckert) [inline protocols/luci-proto-openfortivpn/htdocs/luci-static/resources/protocol/openfortivpn.js:150]: Missing space after comma of the forth argument

File: protocols/luci-proto-openfortivpn/htdocs/luci-static/resources/protocol/openfortivpn.js [hunk 3]
```diff
@@ -37,7 +90,8 @@ return network.registerProtocol('openfortivpn', {
 	},
 
 	renderFormOptions: function(s) {
-		var o;
+		var certLoadPromise = null,
```
Comment (feckert) [inline protocols/luci-proto-openfortivpn/htdocs/luci-static/resources/protocol/openfortivpn.js:93]: Why is this a global variable

File: protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	   type(args.interface) ~= "string" or
+	   not args.interface:match("^[a-zA-Z0-9_]+$")
+	then
+		print(json.stringify({ error = "Invalid interface name" }))
+		os.exit(1)
+	end
+
+	local user_cert_pem = string.format("/etc/openfortivpn/user-cert-%s.pem", args.interface)
+	local user_key_pem = string.format("/etc/openfortivpn/user-key-%s.pem", args.interface)
+	local ca_file_pem = string.format("/etc/openfortivpn/ca-%s.pem", args.interface)
```
Comment (feckert) [inline protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn:62]: Should we not validate if this file does exist?

File: protocols/luci-proto-openfortivpn/root/usr/share/rpcd/acl.d/luci-openfortivpn.json [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+			"ubus": {
+				"luci.openfortivpn": [ "getCertificates" ]
+			},
+			"uci": [ "network" ]
+		},
+		"write": {
+			"ubus": {
+				"luci.openfortivpn": [ "setCertificates" ]
+			},
+			"uci": [ "network" ]
```
Comment (feckert) [inline protocols/luci-proto-openfortivpn/root/usr/share/rpcd/acl.d/luci-openfortivpn.json:14]: Do we need the network read and write permission? Is this not already granted?

File: protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn (hunk 4 above)
Comment (clayface) [reply protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn:66]: Have added an stat check to `readfile`. Return null if it doesn't exist?

**Comment by Neustradamus** [comment]
There is a progress on this PR?

File: protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+if arg[1] == "list" then
+	print(json.stringify({
+		getCertificates = {
+			interface = "interface"
+		},
+		setCertificates = {
+			interface = "interface",
+			user_cert = "user-cert",
+			user_key = "user-key",
+			ca_file = "ca-file"
```
Comment (feckert) [inline protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn:50]: The rpcd variables are not yet correct! https://github.com/openwrt/luci/pull/5448/commits/b1d8328252137906bc1f04e6f58fcce210eea129#diff-edf2506e6803ceae7afb8f2915c5a7d056c9d21020346f52248602ef3382fa31R17

I have now printed out everything again with `ubus -v list` and it seems as if `_` is usual in Openwrt for separating the variables name. So I would suggest that we use this separator. So we have **user_cert**, **user_key** and **ca_file**.
I know I said we should use `-'. Sorry for the inconvenience.

**Review by feckert — CHANGES_REQUESTED** [review]

File: protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn (hunk 6 above)
Comment (clayface) [reply protocols/luci-proto-openfortivpn/root/usr/libexec/rpcd/luci.openfortivpn]: Have made this change. Thanks!

**Comment by feckert** [comment]
Lets try it.
Thanks merged!

**Pull request merged**